### PR TITLE
Autoconditions for flagship (base) attributes

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #7795 (some of it, at least.)

## Feature Details
Adds 2 autocondition prefixes:
`"flagship attribute: <attribute>"` - returns 1000 times the value of the corresponding attribute on the flagship (including all installed outfits).
`"flagship base attribute: <attribute>"` - returns 1000 times the value of the corresponding attribute from the flagship's base attributes (the attributes in the `attributes` section of the ship definition, not including any installed outfits.)

Note that the attributes are multiplied by 1000 before conversion to conditions because conditions only support integers and many ship attributes can be non-integers. `cost` is an exception because the underlying value is stored as an integer already.

### Automated Tests Added
Load a save file with a Peregrine as the flagship and check all the values of the base attribute conditions match what is defined by the ship and check that a few `"flagship attribute: "` values match what I expect given its installed outfits.

## Performance Impact
N/A

### CODE REVIEW

**Comment by Zitchas** [comment]
Awesome. Thank you!

**Comment by Hurleveur** [comment]
Looks good, could you make the integration test do an apply of all those conditions too, so that we know what they are set to if this test fails?

**Review by Hurleveur — APPROVED** [review]
Code wise this is very simple, and the new test tests this pretty extensively, so looks good for me.

**Comment by warp-core** [comment]
Something to note: attributes can be any real number, but conditions can only take integer values.

**Comment by Zitchas** [comment]
> Something to note: attributes can be any real number, but conditions can only take integer values.

That's good to know. So what happens when we have something, such as mass, for instance, that is 100.5. Does the condition round it off?

**Comment by petervdmeer** [comment]
The test appears to be failing in CI, but skipped due to invalid characters
```
Error: Unsupported character(s) in test name:
file ../..//tests/integration/config/plugins/integration-tests/data/tests/tests_flagship_(base)_attribute_autocondition.txt
L112:   test "Flagship (Base) Attribute Autocondition"

1..28
# Running test "Afterburner-flight - Simple depart land":
#     Error: Unsupported character(s) in test name:
#     file /tmp/tmp.Djg6uJPY60/plugins/integration-tests/data/tests/tests_flagship_(base)_attribute_autocondition.txt
#     L112:   test "Flagship (Base) Attribute Autocondition"
ok 1 Afterburner-flight - Simple depart land
# Running test "Atrocity Test - New Atrocity":
#     Error: Unsupported character(s) in test name:
#     file /tmp/tmp.jfIMpx97O3/plugins/integration-tests/data/tests/tests_flagship_(base)_attribute_autocondition.txt
#     L112:   test "Flagship (Base) Attribute Autocondition"
```

(We should probably do something to ensure that invalid tests just cause a failure as well.)

**Comment by warp-core** [comment]
> > Something to note: attributes can be any real number, but conditions can only take integer values.
> 
> That's good to know. So what happens when we have something, such as mass, for instance, that is 100.5. Does the condition round it off?

It looks like non-integer values are always rounded down. (At least, positive ones.)
Should probably add a different method for `has` since anything non-zero should be true for that but the current implementation will have anything between 0 and 1 return 0.

**Comment by Hurleveur** [comment]
Oh wow I just assumed something like a wrong name would fail the tests.

**Review by Amazinite — COMMENTED** [review]
Mass and cost of a ship/outfit aren't attributes in the normal sense, but I'd like to still be able to get that information using these prefixed conditions.

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -2779,6 +2779,35 @@ void PlayerInfo::RegisterDerivedConditions()
...
+	auto flagshipAttributeHelper = [](const Ship *flagship, const string &attribute, bool base) -> int64_t
+	{
+		if(!flagship)
+			return 0;
+
+		const Outfit &attributes = base ? flagship->BaseAttributes() : flagship->Attributes();
+		if(attribute == "cost")
+			return attributes.Cost();
+		if(attribute == "mass")
+			return attributes.Mass() * 1000.;
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:2791]: Why no rounding here?

Comment (warp-core) [reply source/PlayerInfo.cpp:2791]: Because I forgot.

**Comment by Zitchas** [comment]
Question: Can we use conditions in conversations? And by that, I mean use the value in the condition directly. This might be the correct way to do it, but I'm curious as to if there' some way that we could do it:

ex. 'Your ship has "flagship attribute: fuel" fuel. Hopefully that's enough'

**Comment by warp-core** [comment]
> Question: Can we use conditions in conversations? And by that, I mean use the value in the condition directly. This might be the correct way to do it, but I'm curious as to if there' some way that we could do it:
> 
> ex. 'Your ship has "flagship attribute: fuel" fuel. Hopefully that's enough'

No

**Comment by Hurleveur** [comment]
We could however make a real <fuel> text replacement for it
That said you can use substitutions/branching and the amount of fuel to react to it

**Comment by Zitchas** [comment]
I didn't actually expect I could use the condition itself as a text replacement, but I was hoping that we could do something like using a substitution or custom text replacement to pull the value from the attribute. I don't really care about the mechanism behind the scenes, just wanting something that lets me, say, talk about how much shields someone has, or how much cargo space, or stuff like that.

**Pull request merged**